### PR TITLE
Bluetooth: L2CAP: Elevate security level

### LINKED ISSUE #27220: Bluetooth: L2CAP: l2cap_change_security() not considering bt_conn::sec_level when handling BT_L2CAP_LE_ERR_AUTHENTICATION
**Describe the bug**
The client's handling of a `BT_L2CAP_LE_ERR_AUTHENTICATION`, 
when establishing a dynamic channel,
does not take into consideration the link's security level.

**To Reproduce**
1. Create an L2CAP server and set the required minimum security level to `BT_SECURITY_L3`
2. Create an L2CAP client which has no a priori knowledge of the server's required minimum security level
3. Set `CONFIG_BT_DEBUG_L2CAP=y` and observe the logs
4. Connect the peers and set the link's security level to `BT_SECURITY_L2`
5. Ask the client to `bt_l2cap_chan_connect()`
6. Observe that the server is correctly identifying the mismatch between the link's security level and the required minimum security level and thus is refusing the connection with "Insufficient Authentication"
7. Observe that the client is neither trying to "bump" the link's security level nor invoking any `bt_l2cap_chan_ops` callback

**Expected behavior**
The client should correctly identify that given an `BT_L2CAP_LE_ERR_AUTHENTICATION` on a link with `BT_SECURITY_L2`,
it should attempt to "bump" to `BT_SECURITY_L3`.

Depending on the IO capabilities of the devices,
the application on the client side should be informed about the outcome of  `bt_l2cap_chan_connect()`
with either the `bt_l2cap_chan_ops::connected()` or `bt_l2cap_chan_ops::disconnected()` callback.

**Impact**
Any application which uses the client API for an L2CAP dynamic channel will have to workaround this issue,
if it has no a priori knowledge about a peer's L2CAP server required minimum security level,
to avoid stalling the business logic.

**Environment:**
 Commit SHA or Version used: v2.3.0

**Bluetooth Qualification testcase:**
L2CAP/LE/CFC/BV-11-C

### PR DESCRIPTION
**Pull request by joerchan**
Elevate the security level of the connection when the l2cap channel
connect request is denied with encryption or authentication error
code.
The l2cap channel required_sec_level was used, but this is not
correct as this is set by the user of the API. This resulted in
the call to bt_conn_set_security seeing that the current security
level was already set, decide there is nothing to be done and return
success.

Fixes: #27220

Elevate the security level before sending the connection request if
the application has set a required security level on the channel.

### CODE REVIEW

File: subsys/bluetooth/host/att.c [hunk 1]
```diff
@@ -588,7 +588,7 @@ static uint8_t att_handle_rsp(struct bt_att_chan *chan, void *pdu, uint16_t len,
 	k_delayed_work_cancel(&chan->timeout_work);
 
 	if (!chan->req) {
-		BT_WARN("No pending ATT request");
+		BT_DBG("No pending ATT request");
```
Comment (Vudentz) [inline subsys/bluetooth/host/att.c:591]: Under normal situation there is something wrong when the remote respond when there is no request pending, so I wonder if the call path for security failing needs some changes in order not to trigger this.

**Review by Vudentz — COMMENTED** [review]
The descriptions don't seem correct since bt_conn_set_security does not check for chan.required_sec_level, I guess you meant to say that the resulting sec variable would not be correct if the user has set the required_sec_level, I don't think we had introduce required_sec_level to be set by users though so this breaks with the introduction of that not the other way around.

File: subsys/bluetooth/host/smp.c [hunk 2]
```diff
@@ -2778,9 +2779,9 @@ bool bt_smp_request_ltk(struct bt_conn *conn, uint64_t rand, uint16_t ediv, uint
 		/* Notify higher level that security failed if security was
 		 * initiated by slave.
 		 */
-		bt_conn_security_changed(smp->chan.chan.conn,
+		bt_l2cap_encrypt_change(conn, BT_HCI_ERR_PIN_OR_KEY_MISSING);
+		bt_conn_security_changed(conn,
```
Comment (Vudentz) [inline subsys/bluetooth/host/smp.c:2783]: Should bt_conn_security_changed be calling bt_l2cap_encrypt_change? We might as well update the names to be consistent e.g. bt_l2cap_security_changed.

File: subsys/bluetooth/host/l2cap.c [hunk 3]
```diff
@@ -2361,6 +2361,21 @@ static int l2cap_le_connect(struct bt_conn *conn, struct bt_l2cap_le_chan *ch,
...
+		if (err) {
+			return err;
+		}
+
+		atomic_set_bit(ch->chan.status,
+			       BT_L2CAP_STATUS_ENCRYPT_PENDING);
+
+		return 0;
+	}
+
```
Comment (Vudentz) [inline subsys/bluetooth/host/l2cap.c:2387]: Im fine with this change, though I didn't really consider CONFIG_BT_CONN_DISABLE_SECURITY to be taken into account for client APIs just for server.

File: subsys/bluetooth/host/att.c (hunk 1 above)
Comment (joerchan) [reply subsys/bluetooth/host/att.c:591]: I considered this also:
```
static void bt_att_encrypt_change(struct bt_l2cap_chan *chan,
				  uint8_t hci_status)
...
	if (hci_status) {
		if (att_chan->req) {
			att_handle_rsp(att_chan, NULL, 0,
				       BT_ATT_ERR_AUTHENTICATION);
		}

		return;
	}
```
That would mean we wouldn't call `att_process` in this case, that should be fine, right?

File: subsys/bluetooth/host/l2cap.c (hunk 3 above)
Comment (joerchan) [reply subsys/bluetooth/host/l2cap.c:2387]: Ok. I'll remove CONFIG_BT_CONN_DISABLE_SECURITY here then.

File: subsys/bluetooth/host/smp.c (hunk 2 above)
Comment (joerchan) [reply subsys/bluetooth/host/smp.c:2783]: Good idea.

**Comment by joerchan** [comment]
> The descriptions don't seem correct since bt_conn_set_security does not check for chan.required_sec_level, I guess you meant to say that the resulting sec variable would not be correct if the user has set the required_sec_level,

No, the user has not set a required_sec_level, this is initialized to be L1 in `bt_l2cap_chan_connect` in this case.
The connection has been changed to L2, and the remote l2cap server has set sec_level L3.
So when we get AUTHENTICATION_ERR l2cap_change_security used chan->required_sec_level, raised it from L1 to L2, and called bt_conn_set_security with `L2`, since the conn->sec_level is already L2, there is nothing to. Function returns success, and will not signal security changed.

> I don't think we had introduce required_sec_level to be set by users though so this breaks with the introduction of that not the other way around.

That is how it is done for `l2cap_br_conn_security`, and used by avdtp service. I assumed that it should be like this for L2CAP LE CoC as well.
The struct variable is not documented in the header file though.

**Review by jhedberg — APPROVED** [review]

File: subsys/bluetooth/host/att.c (hunk 1 above)
Comment (Vudentz) [reply subsys/bluetooth/host/att.c]: Yep, Im fine with doing the check for pending req.

Comment (joerchan) [reply subsys/bluetooth/host/att.c]: The check has been added.

**Review by Vudentz — APPROVED** [review]
LGTM

**Pull request merged**